src/main/java/org/apache/commons/imaging/formats/png/PngWriter.java

```diff
@@ -29,12 +31,126 @@
+ */
 class PngWriter {
+    
+    /**
+     * Transforms raw image pixel data into the required PNG IDAT
+     * (compressed) form.
+     * 
+     * {@code DataTransformer} can be used to split compressed pixel
+     * data into different chunks.
+     */
+    static final class ChunkedCompressor {
+        
+        /**
+         * Transformation is done on this pixel data.
+         * 
+         * Can be set using {@code setInput} or passed during object
+         * construction.
+         */
+        private byte[] pixelInput;
+        
+        /**
+         * Keeps the amount of data that has been transformed already.
+         */
+        private int transformedTill;
+        
+        private ByteArrayOutputStream outputStream;
+        
+        private DeflaterOutputStream deflaterOutputStream;
+        
+        ChunkedCompressor() {
+            this.pixelInput = null;
+        }
+        
+        ChunkedCompressor(byte[] pixelInput) {
+            setInput(pixelInput);
+        }
+        
+        void setInput(byte[] pixelInput) {
+            this.pixelInput = pixelInput;
+            this.transformedTill = 0;
+            this.outputStream = new ByteArrayOutputStream();
+            this.deflaterOutputStream = new DeflaterOutputStream(outputStream);
+        }
+        
+        /**
+         * Returns the index till which input has been transformed
+         * into output.
+         * 
+         * @return size of transformed input minus 1 
+         */
+        int transformedAmount() {
+            return transformedTill;
+        }
+        
+        /**
+         * Transforms (at maximum) {@code transformLength} bytes of
+         * input into the required output.
+         * 
+         * See {@code writePixelData} for example usage of this
+         * method.
+         * 
+         * @param transformLength
+         * @return 
+         */
+        byte[] transform(int transformLength) throws IOException, ImageWriteException {
+            if(pixelInput == null) {
+                throw new ImageWriteException("Couldn't read pixelInput as it is "
+                        + "null (it wasn't initialized properly, source bug)");
+            }
+            
+            final int end = Math.min(pixelInput.length,
+                    transformedTill + transformLength);
+            final int length = end - transformedTill;
+
+            deflaterOutputStream.write(pixelInput, transformedTill, length);
+            deflaterOutputStream.flush();
+            outputStream.flush();
+
+            final byte[] compressed = outputStream.toByteArray();
+            outputStream.reset();
+            
+            transformedTill += transformLength;
+            return compressed;
+        }
+        
+        /**
+         * Finishes the transformation and closes the deflation
+         * stream.
+         * 
+         * @return
+         * @throws IOException 
+         */
+        byte[] finish() throws IOException {
```

Comment:
I wonder if there is not a new risk introduced here, of streams not being closed. Have not checked the pull request locally to have a better look yet, but might be worth documenting it in the javadocs if possible.